CA-648 Make TokenStore an instantiable class so that it can be injected as a dependency

wchamber opened the pull request:
Make TokenStore an instantiable class so that it can be injected as a dependency.
This will help us create a fake TokenStore that can be injected for unittests so that they do not rely on Datastore.
(has changes from pull#91 included)

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary. Documentation PRs only need 1 thumb.
- [x] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [x] Test this change deployed correctly and works on dev environment after deployment

Commit:
Add a wrapper around memcache to allow that dependency to be injected.

Also add LocalCacheApi for an in-memory implementation of the add/get memcache API used by Bond. This will help us migrate to a different cache implementation.

Commit:
Add unittests for MemcacheApi by making base class CacheApiTest.

Commit:
First pass abstracting out an API for locked storage of a fence token. Needs documentation, better names, tests for DatastoreLockedStorage. Consider if tests in fence_token_vending_test are duplicate now.

Commit:
Rename to 'FenceTokenStorage' and add tests.

Commit:
Make TokenStore injectable as a dependency.

Commit:
Add unittests for MemcacheApi by making base class CacheApiTest.

wchamber committed:
Fix FakeFenceTokenStorage to match changes to FenceTokenStorage API.

wchamber committed:
Merge branch 'wc-CA-648-datastore-fence-token' into wc-CA-648-token-store

wchamber committed:
Remove duplicate 'cache' in comment.

wchamber committed:
Merge branch 'wc-CA-648' of https://github.com/DataBiosphere/bond into wc-CA-648

wchamber committed:
Merge branch 'wc-CA-648' into wc-CA-648-datastore-fence-token

wchamber committed:
Change condition for ServiceAccountNotUpdatedException to raise that function if there is no expired_at message, i.e. it's still locked, instead of only if the updated account does not have a valid expires_at value.
Before, if the update lock expired during the _wait_for_update loop, we could return the FenceServiceAccount without a received key_json, as the update locked FenceServiceAccount does not have any value for expires_at. Now, the expired update locked value raises the exception.

wchamber committed:
Merge branch 'develop' into wc-CA-648

wchamber committed:
Merge branch 'wc-CA-648' into wc-CA-648-datastore-fence-token

wchamber committed:
Go back to catching all exceptions as failures to acquire the locks. Add some more comments on how transactions can fail even when they succeed...

wchamber committed:
Merge branch 'wc-CA-648-datastore-fence-token' into wc-CA-648-token-store

wchamber commented:
I've changed this pull request to pull into branch `wc-CA-datastore-fence-token` as this change is dependent on that branch pull request (#91). This makes this pull request easier to review.
My hope is that pull #91 finishes into develop, then I change this pull request to go into develop.

gpolumbo-broad commented on bond.py:
```diff
@@ -18,6 +18,7 @@ def __init__(self,
         self.oauth_adapter = oauth_adapter
         self.fence_api = fence_api
         self.sam_api = sam_api
+        self.token_store = token_store
```
Let's rename `token_store` to `refresh_token_store`.  I'm finding myself confused between `TokenStore` and `FenceTokenStorage`.  I know this was a legacy naming convention, and we don't necessarily need to rename the class here, but since we're introducing the variables, I wouldn't mind at least naming those more clearly here.

gpolumbo-broad commented on main.py:
```diff
@@ -90,6 +91,7 @@ def __init__(self, fence_tvm, bond):
 class BondApi(remote.Service):
     def __init__(self):
         cache_api = create_cache_api()
+        token_store = TokenStore()
```
Again with the more precise renaming of `token_store` -> `refresh_token_store`

wchamber committed:
Fix requested changes.
Added some comments and renamed functions s/get_or_create/retrieve and s/_fetch_service_account/_fetch_and_cache_service_account and s/create_fence_service_account_key/build_fence_service_account_key.

wchamber committed:
Merge branch 'wc-CA-648-datastore-fence-token' into wc-CA-648-token-store

wchamber committed:
Rename variables token_store -> refresh_token_store. Add a comment to the TokenStore class and a TODO to rename.

wchamber committed:
Merge branch 'develop' into wc-CA-648-datastore-fence-token

wchamber committed:
Merge branch 'wc-CA-648-datastore-fence-token' into wc-CA-648-token-store

wchamber commented:
I've changed the PR back to merge into 'develop' now that PR #91 has merged to develop.
